src/config.js

```diff
@@ -0,0 +1,1602 @@
+//
+// ```coffee
+// config:
+//   someSetting:
+//     title: 'Setting Magnitude'
+//     description: 'This will affect the blah and the other blah'
+//     type: 'integer'
+//     default: 4
+// ```
+//
+// __Note__: You should strive to be so clear in your naming of the setting that
+// you do not need to specify a title or description!
+//
+// Descriptions allow a subset of
+// [Markdown formatting](https://help.github.com/articles/github-flavored-markdown/).
+// Specifically, you may use the following in configuration setting descriptions:
+//
+// * **bold** - `**bold**`
+// * *italics* - `*italics*`
+// * [links](https://atom.io) - `[links](https://atom.io)`
+// * `code spans` - `\`code spans\``
+// * line breaks - `line breaks<br/>`
+// * ~~strikethrough~~ - `~~strikethrough~~`
+//
+// #### order
+//
+// The settings view orders your settings alphabetically. You can override this
+// ordering with the order key.
+//
+// ```coffee
+// config:
+//   zSetting:
+//     type: 'integer'
+//     default: 4
+//     order: 1
+//   aSetting:
+//     type: 'integer'
+//     default: 4
+//     order: 2
+// ```
+//
+// ## Manipulating values outside your configuration schema
+//
+// It is possible to manipulate(`get`, `set`, `observe` etc) values that do not
+// appear in your configuration schema. For example, if the config schema of the
+// package 'some-package' is
+//
+// ```coffee
+// config:
+// someSetting:
+//   type: 'boolean'
+//   default: false
+// ```
+//
+// You can still do the following
+//
+// ```coffee
+// let otherSetting  = atom.config.get('some-package.otherSetting')
+// atom.config.set('some-package.stillAnotherSetting', otherSetting * 5)
+// ```
+//
+// In other words, if a function asks for a `key-path`, that path doesn't have to
+// be described in the config schema for the package or any package. However, as
+// highlighted in the best practices section, you are advised against doing the
+// above.
+//
+// ## Best practices
+//
+// * Don't depend on (or write to) configuration keys outside of your keypath.
+//
+const schemaEnforcers = {}
+
+class Config {
+  static addSchemaEnforcer (typeName, enforcerFunction) {
+    if (schemaEnforcers[typeName] == null) { schemaEnforcers[typeName] = [] }
+    return schemaEnforcers[typeName].push(enforcerFunction)
+  }
+
+  static addSchemaEnforcers (filters) {
+    for (let typeName in filters) {
+      const functions = filters[typeName]
+      for (let name in functions) {
+        const enforcerFunction = functions[name]
+        this.addSchemaEnforcer(typeName, enforcerFunction)
+      }
+    }
+  }
+
+  static executeSchemaEnforcers (keyPath, value, schema) {
+    let error = null
+    let types = schema.type
+    if (!Array.isArray(types)) { types = [types] }
+    for (let type of types) {
+      try {
+        const enforcerFunctions = schemaEnforcers[type].concat(schemaEnforcers['*'])
+        for (let enforcer of Array.from(enforcerFunctions)) {
```

Comment:
`enforcerFunctions` is the result of `.concat()` so we don't need `Array.from()`